### PR TITLE
Support non-append-only COUNTER sources

### PR DESCRIPTION
**Pull request by umanwizard**
Add a `MAX CARDINALITY` parameter to `COUNTER` sources. If specified, the counter load generator will begin retracting the oldest emitted value for each new value it emits, once it has crossed the max cardinality threshold. For example, if a table contains 1, 2, 3, 4 and `MAX CARDINALITY` is set to 4, then after the next few ticks it will contain 2, 3, 4, 5, then 3, 4, 5, 6, then 4, 5, 6, 7, and so on. 

### Motivation

  * This PR adds a feature that has not yet been specified.

    When load testing, `COUNTER` sources are the easiest way to stream changes into the system: they are very easy to use and it is immediately obvious how many tuples will be generated. The one drawback is that if the source runs for a while, any new computations built on top of it will have to churn through all the existing data to catch up. It is occasionally useful for this to not be true, motivating the present change.
    

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Support `MAX CARDINALITY` option on `COUNTER` sources.

### CODE REVIEW

**Comment by umanwizard** [comment]
I initially considered calling this `MAX SIZE`, which is less of a mouthful than `MAX CARDINALITY`, but I decided not to in order to avoid confusion with the `SIZE` parameter on sources (which controls the cluster size).

File: src/storage-client/src/types/sources.proto [hunk 1]
```diff
@@ -214,7 +214,7 @@ message ProtoPostgresSourcePublicationDetails {
 
 message ProtoLoadGeneratorSourceConnection {
     oneof generator {
-        google.protobuf.Empty counter = 1;
+        ProtoCounterLoadGenerator counter = 1;
```
Comment (mjibson) [inline src/storage-client/src/types/sources.proto:217]: Is it safe to reuse ids?

File: src/storage-client/src/types/sources.rs [hunk 2]
```diff
@@ -2188,7 +2194,7 @@ impl LoadGenerator {
                 }
                 DataEncodingInner::RowCodec(desc)
             }
-            LoadGenerator::Counter => DataEncodingInner::RowCodec(
+            LoadGenerator::Counter { max_cardinality: _ } => DataEncodingInner::RowCodec(
```
Comment (mjibson) [inline src/storage-client/src/types/sources.rs:2197]:
```suggestion
            LoadGenerator::Counter { .. } => DataEncodingInner::RowCodec(
```

Comment (mjibson) [reply src/storage-client/src/types/sources.rs:2197]: Could this and below be `..`? I don't care much, but it would match existing code if so.

File: test/testdrive/load-generator.td [hunk 3]
```diff
@@ -29,3 +29,17 @@ users         subsource      <null>
 # Show that AST of subsource contains REFERENCES option
 > SHOW CREATE SOURCE accounts
 materialize.public.accounts "CREATE SUBSOURCE \"materialize\".\"public\".\"accounts\" (\"id\" \"pg_catalog\".\"int8\" NOT NULL, \"org_id\" \"pg_catalog\".\"int8\" NOT NULL, \"balance\" \"pg_catalog\".\"int8\" NOT NULL, UNIQUE (\"id\")) WITH (REFERENCES = true)"
+
+# Check that non-append-only `COUNTER` sources reach the proper size
+
+> CREATE SOURCE counter FROM LOAD GENERATOR COUNTER (MAX CARDINALITY 47, TICK INTERVAL '0.1s') WITH (SIZE '1')
```
Comment (mjibson) [inline test/testdrive/load-generator.td:35]: This will add 4.7s to this test every run. How about reducing the cardinality and tick interval so that this test doesn't block for so long.

File: src/storage/src/source/generator/counter.rs [hunk 4]
```diff
@@ -22,14 +27,50 @@ impl Generator for Counter {
...
+                    Some(max) => {
+                        if max <= counter {
+                            Some(counter - max + 1)
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
+                };
+                counter += 1;
```
Comment (mjibson) [inline src/storage/src/source/generator/counter.rs:43]: If you moved this above to_retract you could remove the +1 in the to_retract calculation.

File: src/storage/src/source/generator/counter.rs [hunk 5]
```diff
@@ -22,14 +27,50 @@ impl Generator for Counter {
...
-                1,
-            )
-        }))
+        let max_cardinality = self.max_cardinality;
+        Box::new(
+            iter::repeat_with(move || {
+                let to_retract = match max_cardinality {
+                    Some(max) => {
+                        if max <= counter {
+                            Some(counter - max + 1)
```
Comment (mjibson) [inline src/storage/src/source/generator/counter.rs:36]: Ah this line has changed the implied type of counter from i64 to usize. Before we had the nice property that it'd rollover at i64 max, now it panics instead and we have 2 try_into's. It'd be nice to never panic and never call try_into.

What if max_cardinality is an `Option<i64>` instead (contrary to my comment above)? The planner step could do the validation to ensure it's > 0, and then maybe all the math gets very clean?

Oh it'd still have the property that after rollever the max cardinality stops working because of the `max <= counter` check. That could be fixed by having two counters: `counter` and `until_retraction` that would start at `max_cardinality` and either subtract 1 from itself or if it's 0 then stay at 0 and do the retraction.

Does any of this matter because i64 is still a number high enough we'll never ever see it even at the lowest tick interval possible?

File: src/storage/src/source/generator/counter.rs [hunk 6]
```diff
@@ -13,7 +13,12 @@ use mz_ore::now::NowFn;
 use mz_repr::{Datum, Row};
 use mz_storage_client::types::sources::{Generator, GeneratorMessageType};
 
-pub struct Counter {}
+pub struct Counter {
+    /// How many values will be emitted
+    /// before old ones are retracted, or `None` for
+    /// an append-only collection.
+    pub max_cardinality: Option<usize>,
```
Comment (mjibson) [inline src/storage/src/source/generator/counter.rs:20]: Could this easily be a u64 instead? That'd match the proto stuff. Or does it make the math below harder?

**Review by mjibson — COMMENTED** [review]
For such a simple sounding feature the types on this get annoying real quick.

File: src/storage-client/src/types/sources.proto (hunk 1 above)
Comment (umanwizard) [reply src/storage-client/src/types/sources.proto:217]: https://materializeinc.slack.com/archives/C0246GEHL8N/p1675971029060749?thread_ts=1675971007.308549&cid=C0246GEHL8N assuming this isn't "one of the few that is persisted to disk", though I have no idea.

File: src/storage/src/source/generator/counter.rs (hunk 6 above)
Comment (umanwizard) [reply src/storage/src/source/generator/counter.rs:20]: I made this a `usize` because it felt more semantically correct to me. It has to be converted to `u64` on the boundaries because protobuf doesn't have the concept of `usize`. I could also make it `u64`; I don't feel strongly.

File: src/storage/src/source/generator/counter.rs (hunk 5 above)
Comment (umanwizard) [reply src/storage/src/source/generator/counter.rs:36]: The lowest tick interval is 1µs, which will take hundreds of thousands of years to overflow an i64. Given that fact, I don't think it makes sense to make the code more complicated to support the overflow case.

Comment (mjibson) [reply src/storage/src/source/generator/counter.rs:36]: I think I'm ok not doing all of ^ if it's really annoying as long as we add a comment saying it'll never panic in practice because even at a tick interval of 1ns it'll still take X millennia before it triggers or something.

Hmm what if we ever want to support a tick interval of 0 to simulate a busy source? Maybe it should be safer.

File: src/storage/src/source/generator/counter.rs (hunk 4 above)
Comment (umanwizard) [reply src/storage/src/source/generator/counter.rs:43]: Yes, but then the if statement would have to read `if max <= counter - 1`, which doesn't seem like a net win.

File: src/storage/src/source/generator/counter.rs (hunk 6 above)
Comment (umanwizard) [reply src/storage/src/source/generator/counter.rs]: (We discussed that on Slack and it seems like it results in the simplest code)

File: src/storage-client/src/types/sources.proto (hunk 1 above)
Comment (mjibson) [reply src/storage-client/src/types/sources.proto:217]: I also don't know. Can you figure this out, or do the normal reserved thing if that's too hard/slow/annoying?

**Review by mjibson — APPROVED** [review]
LGTM modulo proto stuff.

File: test/testdrive/load-generator.td [hunk 7]
```diff
@@ -29,3 +29,21 @@ users         subsource      <null>
 # Show that AST of subsource contains REFERENCES option
 > SHOW CREATE SOURCE accounts
 materialize.public.accounts "CREATE SUBSOURCE \"materialize\".\"public\".\"accounts\" (\"id\" \"pg_catalog\".\"int8\" NOT NULL, \"org_id\" \"pg_catalog\".\"int8\" NOT NULL, \"balance\" \"pg_catalog\".\"int8\" NOT NULL, UNIQUE (\"id\")) WITH (REFERENCES = true)"
+
+# Check that non-append-only `COUNTER` sources reach the proper size
```
Comment (mjibson) [inline test/testdrive/load-generator.td:33]: Can you add a test for max card 0? Just want to make sure it doesn't panic due to some edge case.

File: src/storage-client/src/types/sources.proto (hunk 1 above)
Comment (umanwizard) [reply src/storage-client/src/types/sources.proto]: Dunno how to do that without auditing the entire codebase, so I went ahead and just reserved the old field number.
